File: pc_nodes_examples/cpp/src/zenoh_test_sub.cpp
#include <zenoh_ros/ZenohRosPC.h>
#include <zenoh_ros/std_msgs/z_Int32.h>

class CounterSubscriberNode : public ZenohNode {
public:
    CounterSubscriberNode() : ZenohNode("counter_subscriber") {
        ZLOG_INFO(this->get_logger(), "Node has been started");

        // Create subscription to standard ROS 2 message type (Int32) and depth 10
        sub_ = this->z_create_subscription<z_Int32>(
            "robot/sim_counter",
            [this](const z_Int32& msg) -> void {
                this->listener_callback(msg);
            },
            10
        );
    }

private:
    ZenohSubscription<z_Int32>* sub_ = nullptr;

    void listener_callback(const z_Int32& msg) {
        ZLOG_INFO(this->get_logger(), "[RECV FROM MCU] sim_counter: %d", msg.data);
    }
};

int main(int argc, char** argv) {
    (void)argc;
    (void)argv;

    // Initialize Zenoh session — connect to MCU SoftAP via Wi-Fi
    ZenohConfig config;
    config.communication_mode = ZenohCommunicationMode::ZENOH_COMM_WIFI;
    config.host = "10.42.0.50";
    config.port = 7447;

    if (!ZenohNode::init(config)) {
        return -1;
    }

    {
        CounterSubscriberNode node;
        node.z_spin();
    } // Node goes out of scope and destroys subscriptions cleanly before session closes via atexit

    return 0;
}

File: pc_nodes_examples/python/zenoh_test_sub.py
#!/usr/bin/env python3
import sys
import os

# Ensure shared zenoh_ros package can be imported
sys.path.append(os.path.abspath(os.path.join(os.path.dirname(__file__), '../../shared_libraries/python')))

from zenoh_ros import ZenohNode, ZenohConfig
from zenoh_ros.std_msgs import z_Int32

class CounterSubscriberNode(ZenohNode):
    def __init__(self) -> None:
        super().__init__("counter_subscriber")
        self.get_logger().info("Node has been started")

        # Create subscription to standard ROS 2 message type (z_Int32) and depth 10
        self.sub = self.z_create_subscription(
            z_Int32,
            "robot/sim_counter",
            self.listener_callback,
            10
        )

    def listener_callback(self, msg: z_Int32) -> None:
        self.get_logger().info(f"[RECV FROM MCU] sim_counter: {msg.data}")


def main() -> None:
    cfg = ZenohConfig(communication_mode=ZenohConfig.ZENOH_COMM_WIFI, host="10.42.0.50", port=7447)
    if not ZenohNode.init(cfg):
        return

    node_instance = None
    try:
        node_instance = CounterSubscriberNode()
        node_instance.z_spin()

    except KeyboardInterrupt:
        if node_instance:
            node_instance.get_logger().info("Shutdown requested via KeyboardInterrupt.")

    except Exception as e:
        if node_instance:
            node_instance.get_logger().error(f"Critical Error: {e}")

    finally:
        ZenohNode.shutdown()



if __name__ == "__main__":
    main()